pkg/admission/validation.go

```diff
@@ -0,0 +1,112 @@
+/*
+ * Tencent is pleased to support the open source community by making TKEStack available.
+ *
+ * Copyright (C) 2012-2019 Tencent. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package admission
+
+import (
+	"strconv"
+
+	tappv1 "tkestack.io/tapp/pkg/apis/tappcontroller/v1"
+
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	validation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+const isZeroErrorMsg string = `must be greater than 0`
+
+func validate(tapp *tappv1.TApp) error {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validateTAppSpec(&tapp.Spec, field.NewPath("spec"))...)
+	if len(allErrs) > 0 {
+		return allErrs.ToAggregate()
+	}
+	return nil
+}
+
+func validateTAppSpec(spec *tappv1.TAppSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, validateTAppUpdateStrategy(&spec.UpdateStrategy, fldPath.Child("updateStrategy"))...)
+
+	return allErrs
+}
+
+func validateTAppUpdateStrategy(strategy *tappv1.TAppUpdateStrategy, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if strategy.MaxUnavailable != nil {
+		allErrs = append(allErrs, validatePositiveIntOrPercent(*strategy.MaxUnavailable, fldPath.Child("maxUnavailable"))...)
```

Comment:
Would it be better to put all these validation to one function, e.g. `validateIntOrPercent`? line 53\~55 and 58\~60 are almost same, they can call a same function.